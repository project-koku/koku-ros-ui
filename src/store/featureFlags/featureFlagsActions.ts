import { createAction } from 'typesafe-actions';

export interface FeatureFlagsActionMeta {
  isExportsFeatureEnabled?: boolean;
  isFinsightsFeatureEnabled?: boolean;
  isIbmFeatureEnabled?: boolean;
  isRosFeatureEnabled?: boolean;
  isSettingsPlatformFeatureEnabled?: boolean;
}

export const setFeatureFlags = createAction('feature/init_feature_flags')<FeatureFlagsActionMeta>();
export const resetState = createAction('feature/reset_state')();
