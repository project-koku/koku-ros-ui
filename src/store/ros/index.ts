import * as rosActions from './rosActions';
import { rosStateKey } from './rosCommon';
import type { CachedRos, RosAction, RosState } from './rosReducer';
import { rosReducer } from './rosReducer';
import * as rosSelectors from './rosSelectors';

export type { RosAction, CachedRos, RosState };
export { rosActions, rosReducer, rosSelectors, rosStateKey };
