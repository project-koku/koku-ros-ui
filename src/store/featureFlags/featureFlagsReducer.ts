import type { ActionType } from 'typesafe-actions';
import { getType } from 'typesafe-actions';

import type { resetState } from './featureFlagsActions';
import { setFeatureFlags } from './featureFlagsActions';

export type FeatureFlagsAction = ActionType<typeof setFeatureFlags | typeof resetState>;

export type FeatureFlagsState = Readonly<{
  hasFeatureFlags: boolean;
  isExportsFeatureEnabled: boolean;
  isFinsightsFeatureEnabled: boolean;
  isIbmFeatureEnabled: boolean;
  isRosFeatureEnabled: boolean;
  isSettingsPlatformFeatureEnabled: boolean;
}>;

export const defaultState: FeatureFlagsState = {
  hasFeatureFlags: false,
  isExportsFeatureEnabled: false,
  isFinsightsFeatureEnabled: false,
  isIbmFeatureEnabled: false,
  isRosFeatureEnabled: false,
  isSettingsPlatformFeatureEnabled: false,
};

export const stateKey = 'featureFlags';

export function featureFlagsReducer(state = defaultState, action: FeatureFlagsAction): FeatureFlagsState {
  switch (action.type) {
    case getType(setFeatureFlags):
      return {
        ...state,
        hasFeatureFlags: true,
        isExportsFeatureEnabled: action.payload.isExportsFeatureEnabled,
        isFinsightsFeatureEnabled: action.payload.isFinsightsFeatureEnabled,
        isIbmFeatureEnabled: action.payload.isIbmFeatureEnabled,
        isRosFeatureEnabled: action.payload.isRosFeatureEnabled,
        isSettingsPlatformFeatureEnabled: action.payload.isSettingsPlatformFeatureEnabled,
      };

    default:
      return state;
  }
}
