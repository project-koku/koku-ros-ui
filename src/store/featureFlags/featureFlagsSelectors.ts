import type { RootState } from 'store/rootReducer';

import { stateKey } from './featureFlagsReducer';

export const selectFeatureFlagsState = (state: RootState) => state[stateKey];

export const selectHasFeatureFlags = (state: RootState) => selectFeatureFlagsState(state).hasFeatureFlags;

export const selectIsExportsFeatureEnabled = (state: RootState) =>
  selectFeatureFlagsState(state).isExportsFeatureEnabled;
export const selectIsFinsightsFeatureEnabled = (state: RootState) =>
  selectFeatureFlagsState(state).isFinsightsFeatureEnabled;
export const selectIsIbmFeatureEnabled = (state: RootState) => selectFeatureFlagsState(state).isIbmFeatureEnabled;
export const selectIsRosFeatureEnabled = (state: RootState) => selectFeatureFlagsState(state).isRosFeatureEnabled;
export const selectIsSettingsPlatformFeatureEnabled = (state: RootState) =>
  selectFeatureFlagsState(state).isSettingsPlatformFeatureEnabled;
