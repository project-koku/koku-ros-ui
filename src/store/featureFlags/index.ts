import * as featureFlagsActions from './featureFlagsActions';
import type { FeatureFlagsAction, FeatureFlagsState } from './featureFlagsReducer';
import { featureFlagsReducer, stateKey as featureFlagsStateKey } from './featureFlagsReducer';
import * as featureFlagsSelectors from './featureFlagsSelectors';

export type { FeatureFlagsAction, FeatureFlagsState };
export { featureFlagsActions, featureFlagsReducer, featureFlagsSelectors, featureFlagsStateKey };
