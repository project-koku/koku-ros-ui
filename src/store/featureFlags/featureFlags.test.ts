import { createMockStoreCreator } from 'store/mockStore';

import { featureFlagsSelectors } from '.';
import * as actions from './featureFlagsActions';
import { featureFlagsReducer, stateKey } from './featureFlagsReducer';
import * as selectors from './featureFlagsSelectors';

const createUIStore = createMockStoreCreator({
  [stateKey]: featureFlagsReducer,
});

test('default state', async () => {
  const store = createUIStore();
  expect(selectors.selectFeatureFlagsState(store.getState())).toMatchSnapshot();
});

test('Exports feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureFlags({ isExportsFeatureEnabled: true }));
  expect(featureFlagsSelectors.selectIsExportsFeatureEnabled(store.getState())).toBe(true);
});

test('FINsights feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureFlags({ isFinsightsFeatureEnabled: true }));
  expect(featureFlagsSelectors.selectIsFinsightsFeatureEnabled(store.getState())).toBe(true);
});

test('IBM feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureFlags({ isIbmFeatureEnabled: true }));
  expect(featureFlagsSelectors.selectIsIbmFeatureEnabled(store.getState())).toBe(true);
});

test('ROS feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureFlags({ isRosFeatureEnabled: true }));
  expect(featureFlagsSelectors.selectIsRosFeatureEnabled(store.getState())).toBe(true);
});
