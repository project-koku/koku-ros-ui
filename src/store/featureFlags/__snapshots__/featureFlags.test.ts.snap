// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default state 1`] = `
{
  "hasFeatureFlags": false,
  "isExportsFeatureEnabled": false,
  "isFinsightsFeatureEnabled": false,
  "isIbmFeatureEnabled": false,
  "isRosFeatureEnabled": false,
  "isSettingsPlatformFeatureEnabled": false,
}
`;
